### PR TITLE
Update to upstream pthread

### LINKED ISSUE #26: can we use upstream pthread instead of this library?
The new `newlib` update from devkitPro seems to include an actual `libpthread`!!!
```console
$ arm-none-eabi-nm   /opt/devkitpro/devkitARM/arm-none-eabi/lib/libsysbase.a | grep pthread | wc -l 
86
```

I can't quite tell if it works correctly or at all, but there is definitely some new code in the `libsysbase` library and we might need to readjust how this library works to play nicely with it (link errors etc.)

The implementation looks to be similar to what we have done in this library, but perhaps the syscalls are still stubs (and we could implement them here, but let `newlib` handle the business logic):

https://github.com/devkitPro/buildscripts/blob/cd5e224d6a14f7d32712ab10cfc08e0c6a2daea3/dkarm-eabi/patches/newlib-4.3.0.20230120.patch#L7741

There might also be some other stuff that relates to `shim-3ds`, it seems like a lot of new library surface area. Probably will take a while to crawl through all of it and figure out what we can use or need to change.

### PR DESCRIPTION
**Pull request by Meziu**
Closes #26 

This is an update to base our implementation on top of the upstread `pthread` that has been recently added.
Most calls have been either removed (if they were fully re-implemented by `newlib`, like those for Mutexes and Condvars) or have been morphed into "syscalls" (as they call them), which are downstream implementations of the pthread functions.

There are a few things to note here:
- The package size decreased immensely because a big chunk of the implementations is already implemented.
- All of `libc` seems to be already compatible with the changes, so there are no changes we *need* to do over there.
- The new `__syscall_thread_create` function _*CANNOT*_ receive priority or cpu affinity as its parameters, since it doesn't receive a `pthread_attr_t` as argument (very infuriating)

The last part is absolutely horrendous, since it would mean a BIG step back on our progress to "normalize" priority and affinity. However, we can't ignore the massive importance of these new impls, and that probably `libctru` itself will one day support them directly.

As such, I made this PR to show you what a "new" `pthread-3ds` may look like, and to discuss its future.

I'll start the discussion on 2 questions:
1. If we can set the priority when a thread is *already* running, maybe we don't need to pass the priority during creation?
2. Since, to support cpu affinity (especially on the sys_core), we would need a custom thread creation process regardless, can we simply avoid passing the special arguments to `pthread-3ds` and instead use a different solution? (something like a temporary thread local variable which we can set from `ctru-rs` and read in `pthread-3ds`).

I know these solutions are pretty ugly and unstable, but I honestly don't think we can stop the changes made in `newlib` or `libctru` simply because "we are against that". There is progress being made on those fronts and perhaps we should just adapt. However, we should try to PR some changes regarding to “sched_yield” and “thread_create”, because as it stands we just lose functionality.

Edit: This PR has been thoroughly tested with `std::thread::spawn` and `std::sync::Mutex`, so I am 100% certain it works. However, it requires the latest dependencies from dkARM and [this](https://github.com/rust3ds/ctru-rs/pull/110) `ctru-sys`.

### CODE REVIEW

File: src/misc.rs [hunk 1]
```diff
@@ -1,26 +1,13 @@
 //! Miscellaneous pthread functions
 
+// The implementation within `newlib` stubs this out *entirely*. It's not possible to use a "syscall".
```
Comment (ian-h-chamberlain) [inline src/misc.rs:3]: Looking upstream, I see `int __attribute__((weak)) sched_yield (void)` - so we should be able to override this with our version, right?

**Comment by ian-h-chamberlain** [comment]
> The new `__syscall_thread_create` function __CANNOT__ receive priority or cpu affinity as its parameters, since it doesn't receive a `pthread_attr_t` as argument (very infurianting)

Ugh, this is really unfortunate, especially since the `pthread_create` etc. are not weakly linked so we can't even override those implementations. I wonder if we can set priority / affinity just after thread creation instead? [One of the comments](https://github.com/rust-lang/libs-team/issues/195#issuecomment-1485542010) on the new proposal I opened suggested an API like that on the `Builder`, so maybe that would be a good direction for us to go, if we can't upstream changes to `newlib`?

**Comment by ian-h-chamberlain** [comment]
> Is there any way to test if we can actually change the priority during runtime? We already have an [implementation](https://github.com/rust3ds/pthread-3ds/blob/c885d8cda6d0c5b429b5db3a2408bf204fc1097e/src/thread.rs#L318) of `pthread_setschedparam`, but I remember that the OS specifics (around `svcSetThreadPriority`) are still a bit undocumented.
> ...
> I could try to run some tests, but I'm very low on time.

I can try to do some testing today, and see what all the various functions actually do. My guess is that those functions wouldn't exist if they didn't do something, but it's also possible they are just backports or something from Switch (or they only work in kernelspace and not userspace, require privileged app etc).

**Comment by Meziu** [comment]
Yeah, after some rudimentary tests I got to the same conclusions. Priority can and should be handled by `std`. The callback model seems to work fine so that's not a concern.

About the processorId: the first thing we should do is try to contact (via an issue or whatever) dkArm. If we manage to change the `syscall` signature we are back on having our original functionality. Otherwise, we should look in ways to circumvent the walls via `ctru-rs`, but that's our last solution as it would involve some pretty ugly and unstable code.

Regardless, I wasn't able to find the changes to `pthread` in their [`devkitARM`](https://github.com/devkitPro/newlib/tree/devkitARM) branch, did they publish them anywhere?

**Comment by ian-h-chamberlain** [comment]
> Regardless, I wasn't able to find the changes to `pthread` in their [`devkitARM`](https://github.com/devkitPro/newlib/tree/devkitARM?rgh-link-date=2023-04-13T18%3A20%3A44Z) branch, did they publish them anywhere?

I have been looking into that too, the only place I can find existence of it is https://github.com/devkitPro/buildscripts/blob/cd5e224d6a14f7d32712ab10cfc08e0c6a2daea3/dkarm-eabi/patches/newlib-4.3.0.20230120.patch so far, but perhaps the intention is deliver the same patch to https://github.com/devkitPro/newlib/tree/devkitARM eventually (I see an older commit from when they updated to 4.2.0, and the patch to `buildscripts` is newer than any commit on the `newlib`/`devkitARM` branch).

I think probably what we should do is open an issue against https://github.com/devkitPro/newlib, asking what the general status / expectations of pthreads on 3DS/devkitARM are, without asking for anything just yet. Depending on how that goes we can try to figure out the best path forward, but I don't want to rush into asking for features from the devkitPro team especially since we are trying to do some pretty nonstandard / unsupported stuff as far as they are concerned.

**Comment by Meziu** [comment]
> I think probably what we should do is open an issue against https://github.com/devkitPro/newlib, asking what the general status / expectations of pthreads on 3DS/devkitARM are, without asking for anything just yet. Depending on how that goes we can try to figure out the best path forward, but I don't want to rush into asking for features from the devkitPro team especially since we are trying to do some pretty nonstandard / unsupported stuff as far as they are concerned.

Can I leave the diplomacy to you? You seem pretty experienced on this side of things, lol.

**Comment by ian-h-chamberlain** [comment]
> Can I leave the diplomacy to you? You seem pretty experienced on this side of things, lol.

Haha, sure. I don't know that I'm *so* experienced, just know they have discouraged people from trying to use unsupported languages etc. in the past. I'll try to write something up in the next few days.

**Comment by Meziu** [comment]
> Haha, sure. I don't know that I'm _so_ experienced, just know they have discouraged people from trying to use unsupported languages etc. in the past.

They have tried with us already. And yet, it seems we are the ones who got the furthest in actually implementing a language. 😉

**Comment by Meziu** [comment]
@ian-h-chamberlain Well, it seems like this issue got where we predicted. I've just noticed that @fincs has officially started supporting the new signature in `libctru` via this commit https://github.com/devkitPro/libctru/commit/e9fa000e941398838490ebeda032b99a23d23c47. It hasn't made its way into a release yet, but it's only a matter of days.

In their implementation the cpu-id issue got "ignored" (obviously, there isn't much to do about it), and instead they just default to spawning threads on the [main core](https://github.com/devkitPro/libctru/blob/e9fa000e941398838490ebeda032b99a23d23c47/libctru/source/system/syscalls.c#L204). Those changes also invalidate the need for even more functions which are still present in this PR, so my guess is that this repo as a whole will end up exactly like [`shim-3ds`](https://github.com/rust3ds/shim-3ds).

From one point of view it's great, we need to maintain less code thanks to the advancements in the upstream toolchain, which is exactly what I wanted when I wrote this crate. However, unlike `shim-3ds`, where we basically got a free pass on the whole crate, I feel like `pthread` wasn't handled in the best way possible. I immensely respect the effort put into the whole project, but I believe the merge between `dkA64` and `dkARM` made the support for the 3DS a tiny bit less precise. I just wished there was a way to pass `pthread_attr` to the underlying implementation (in a standard way).

At the end of the day, I can't really complain. Our best bet for now is to just adapt and stabilize our tools as much as possible. I believe we should port what remains of this repo over to `ctru-rs` and maybe work on system-specific functionality in there as well.

**Comment by AzureMarker** [comment]
Maybe I missed a thread, but I don't see much follow-up from our side besides this issue:
https://github.com/devkitPro/newlib/issues/26

We can reply again and ask for this feature request to be considered, and/or open an PR ourselves.

**Comment by Meziu** [comment]
> We can reply again and ask for this feature request to be considered, and/or open an PR ourselves.

I was thinking about re-discussing the issue in that thread/possible PRs and to generally work on it, but there is some complexity regarding the cross-support with Nintendo Switch code (which doesn't need our changes but would get impacted by them) and the newly added implementations in `libctru` which is making me desist from getting involved. It's still a very dynamic situation and I'm honestly prioritizing working on our packages while I can (stuff such as the gazillion lines of documentation I'm currently writing for `ctru-rs` after the `cargo-3ds` release).

**Comment by fincs** [comment]
Fwiw, the current impl in libctru is still experimental/unfinished, do not expect it to ship in the near future. As mentioned previously, specifying custom thread priorities/affinity masks is currently not supported. The 3DS's version of the Horizon kernel has some limitations that are not a good match for the pthreads interface, and do not play nice with typical pthread-using code that expects a mainstream Unix-like kernel. For example:

- 3DS Horizon only supports SCHED_FIFO policy, for 0x40 given priority levels. Switch Horizon additionally supports SCHED_OTHER, and only within a single (low) priority value. Most pthreads apps assume SCHED_OTHER is always available and used. [More on scheduling policies](https://man7.org/linux/man-pages/man7/sched.7.html)
- 3DS Horizon does not support automatic core migration at all (i.e. the kernel automatically selecting the best core to run a thread on at a certain time), and threads are always pinned to a given core. Changing the core assigned to a thread after creation does not make much sense either, due to the core layout used on 3DS (see below point). Switch Horizon on the other hand supports core migration, and we currently specify all created threads to be runnable on cores 0,1,2. 
- 3DS Horizon needs to take into account the fact that core1 is used by system processes, and applications can only create a single thread on core1, and only access a limited timeslice of it. On New3DS systems, core2 acts as a second core available to apps (but remember: no core migration), while core3 is generally unavailable to apps.

With this said, we are nonetheless open to improving the configurability of standard threading APIs in order to better exploit Horizon's threading system and facilitate compatibility with ported software. Any suggestions or proposals are of course welcome.

**Comment by ian-h-chamberlain** [comment]
As an update, it looks like an initial implementation has landed in `libctru`: https://github.com/devkitPro/libctru/releases/tag/v2.3.0

It might be a good time to revive this work — we could probably eliminate even more of this library now that the syscalls are implemented upstream in https://github.com/devkitPro/libctru/blob/master/libctru/source/system/syscalls.c

**Comment by Meziu** [comment]
Alright, it happened. I will look into the changes once I get a break from university exams. Thanks for the heads up @ian-h-chamberlain.

**Comment by Meziu** [comment]
After looking into it a bit more, I've chosen not to merge the changes in the current state, since we are able to avoid using `libctru`'s definitions (which still lack functionality compared to our bindings).

If we ever *need* to use them in the future, we should work on those upstream C definitions to regain compatibility, rather than change stuff here.

**Pull request closed without merging**